src/V3EmitC.cpp

```diff
@@ -1337,16 +1377,85 @@ class EmitVarTspSorter final : public V3TSP::TspStateBase {
 
 unsigned EmitVarTspSorter::s_serialNext = 0;
 
+//######################################################################
+// Emit lazy forward declarations
+
+class LazyDecls final : public AstNVisitor {
+    // NODE STATE/TYPES
+    //  AstNode::user2() -> bool. Already emitted decl for symbols.
+    AstUser2InUse m_inuser2;
+
+    // MEMBERS
+    std::unordered_set<string> m_emittedManually;  // Set of names already declared manually.
+    EmitCBaseVisitor& m_emitter;  // For access to file output
+    bool m_needsBlankLine;  // Emit blank line if any declarations are emitted (purely cosmetic)
+
+    void lazyDeclare(AstCFunc* funcp) {
+        if (funcp->user2SetOnce()) return;  // Already declared
+        if (!funcp->isMethod() || !funcp->isLoose()) return;  // Not lazily declared
+        if (m_emittedManually.count(funcp->nameProtect())) return;  // Already declared manually
+        m_emitter.emitCFuncDecl(funcp, VN_CAST_CONST(funcp->user4p(), NodeModule));
+        m_needsBlankLine = true;
+    }
+
+    virtual void visit(AstNodeCCall* nodep) override {
+        lazyDeclare(nodep->funcp());
+        // Assume args don't contain calls themselves, which is true for now, so don't iterate ...
+    }
+
+    virtual void visit(AstAddrOfCFunc* nodep) override {  //
+        lazyDeclare(nodep->funcp());
+    }
+
+    virtual void visit(AstExecGraph* nodep) override {
+        if (nodep->user2SetOnce()) return;  // Already declared
+        // Build the list of initial mtasks to start
+        for (const ExecMTask* mtp : nodep->rootMTasks()) {
+            m_emitter.puts("void ");
+            m_emitter.puts(EmitCBaseVisitor::topClassName() + "__"
+                           + EmitCBaseVisitor::protect(mtp->cFuncName()));
+            m_emitter.puts("(void* voidSelf, bool even_cycle);\n");
+            m_needsBlankLine = true;
+        }
+    }
+
+    virtual void visit(AstNode* nodep) override { iterateChildrenConst(nodep); }
+
+    VL_DEBUG_FUNC;
+
+public:
+    LazyDecls(EmitCBaseVisitor& emitter)
+        : m_emitter(emitter) {}
+    void emit(AstNode* nodep) {
+        m_needsBlankLine = false;
```

Comment:
```suggestion
```

Reply:
Again, this is inside the `emit` function, not the constructor, basically we want a blank line every time `emit` is invoked and actually emits a declaration.